ckan/lib/helpers.py

```diff
@@ -2908,3 +2908,78 @@ def csrf_input():
     '''
     import ckan.lib.base as base
     return literal(base.render('snippets/csrf_input.html'))
+
+
+@core_helper
+def csrf_session_enabled() -> bool:
+    """ Returns true if the session is enabled for CSRF protection """
+    log.debug("WTF_CSRF_FIELD_NAME: %r", config.get('WTF_CSRF_FIELD_NAME'))
+    log.debug("session: %r", session)
+
+    session_accessed = session.accessed
+    is_enabled = config.get('WTF_CSRF_FIELD_NAME') in session
+    # Reset session accessed state, so we don't introduce vary cookie accidentally
+    session.accessed = session_accessed
+    return is_enabled
+
+
+@core_helper
+def cache_level():
+    return getattr(g, 'cache_type', None)
+
+
+@core_helper
+def limit_cache_for_page() -> Optional[bool]:
+    return getattr(g, 'limit_cache_for_page', None)
+
+
+@core_helper
+def set_limit_cache_for_page(limit: bool) -> None:
+    g.limit_cache_for_page = limit
+
+
+@core_helper
+def set_cache_level(cache_type: 'CacheType|str',
+                    force: bool = False) -> Optional[CacheType]:
+    """Allow setting the cache without downgrading cache unless forced
+    force: Use with caution for example, downgrading cache to public
+    when logged in can have major side effects"""
+    if isinstance(cache_type, str):
+        try:
+            cache_type = CacheType(cache_type)
+        except ValueError:
+            log.warning("Invalid Cache Type passed in, received %r. Ignoring",
+                        cache_type)
+            return None
+
+    currentCacheType = cache_level()
+
+    if currentCacheType and cache_type:
+        if CacheType.can_override(currentCacheType, cache_type) or force:
+            g.cache_type = cache_type
+    else:
+        g.cache_type = cache_type
+    # log.debug('cacheType set to %r', cache_type)
+    return g.cache_type
+
+
+@core_helper
+def etag_append(append_value: str) -> None:
+    """ Adds additional etag uniqueness, can be called multiple times"""
+    current_value = getattr(g, 'etag_append', "")
+    g.etag_append = current_value + append_value
```

Comment:
appending is unsafe for collisions, we can use an HMAC-like approach and store the hash after appending so that
```
h.etag_append('Smith')
h.etag_append('anson')
```
will produce a different etag than
```
h.etag_append('Smit')
h.etag_append('hanson')
```

Reply:
The append will add to the current hamc like hash we have. it allows multiple extensions to add their own 'hash' . so if you call 
``h.etag_append('Smith'); h.etag_append('anson')``
then the etag would now have ``{mtime}-{size}-{check}{etag_append} `` and etag_append would be ``Smithanson``

this does not change the etag completely.  That is done by function ``set_etag_replace`` which if set append won't be used.